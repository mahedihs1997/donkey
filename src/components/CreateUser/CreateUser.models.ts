export interface UserValues {
  username: string;
  name: string;
  password: string;
}

export interface CreateUserProps {
  onSuccess: () => void;
  open: boolean;
  handleClose: () => void;
}
