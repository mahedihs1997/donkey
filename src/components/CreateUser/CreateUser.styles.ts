import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles({
  form: {
    maxWidth: 400,
  },
  button: {
    marginTop: 30,
  },
});
