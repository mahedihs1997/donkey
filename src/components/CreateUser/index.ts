export { default } from './CreateUser';
export type { CreateUserRefs } from './CreateUser.models';
