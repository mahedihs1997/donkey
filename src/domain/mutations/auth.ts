import gql from 'graphql-tag';

export const LOGIN_MUTATION = gql`
  mutation Login($username: String!, $password: String!) {
    signIn(login: $username, password: $password) {
      token
    }
  }
`;
