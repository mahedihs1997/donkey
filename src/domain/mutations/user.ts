import gql from 'graphql-tag';

export const CREATE_USER_MUTATION = gql`
  mutation CreateUser($username: String!, $password: String!, $name: String!) {
    createUser(
      data: { username: $username, name: $name, password: $password }
    ) {
      token
    }
  }
`;
