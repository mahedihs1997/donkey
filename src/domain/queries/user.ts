import gql from 'graphql-tag';

export const ME_QUERY = gql`
  query Me {
    me {
      username
      role
    }
  }
`;

export const USERS_LIST_QUERY = gql`
  query Users {
    users {
      username
      name
    }
  }
`;
