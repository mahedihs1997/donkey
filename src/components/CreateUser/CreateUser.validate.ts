import { UserValues } from './CreateUser.models';
import { FormikErrors } from 'formik';

export const validateCreateUser = (values: UserValues) => {
  const errors: FormikErrors<UserValues> = {};
  if (!values.username) {
    errors.username = 'Username is required';
  }

  if (!values.password) {
    errors.password = 'Enter a password';
  }

  if (!values.name) {
    errors.name = 'Enter a name';
  }

  return errors;
};
