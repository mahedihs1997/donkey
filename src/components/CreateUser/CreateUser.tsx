import React from 'react';
import { validateCreateUser } from './CreateUser.validate';
import { toast } from 'react-toastify';
import { CREATE_USER_MUTATION } from '@domain/mutations/user';
import { useMutation } from '@apollo/react-hooks';
import { Formik, Form } from 'formik';
import FormikTextInput from '@components/FormikTextInput';
import Button from '@material-ui/core/Button';
import { CreateUserProps } from './CreateUser.models';
import { useStyles } from './CreateUser.styles';

const CreateUser: React.FunctionComponent<CreateUserProps> = (props) => {
  const { onSuccess } = props;
  const [createUser] = useMutation(CREATE_USER_MUTATION);

  const classes = useStyles();

  const notifyError = () => {
    toast.error('Creating user failed ðŸ˜‹');
  };

  const notifySuccess = () => {
    toast.success('Creating user succeeded ðŸ˜‹');
  };

  return (
    <Formik
      initialValues={{ name: '', password: '', username: '' }}
      validate={validateCreateUser}
      onSubmit={async (values, actions) => {
        try {
          await createUser({
            variables: {
              ...values,
            },
          });

          notifySuccess();
          onSuccess();
          actions.resetForm();
        } catch (e) {
          notifyError();
          console.error('Creating user failed with error: ', e);
        }
      }}
    >
      {(props) => (
        <Form>
          <div className={classes.form}>
            <FormikTextInput name="name" label="Name" />
            <FormikTextInput name="username" label="Username" />
            <FormikTextInput name="password" label="Password" />
            <Button
              variant="contained"
              color="primary"
              type="submit"
              disabled={props.isSubmitting}
              className={classes.button}
            >
              Create New User
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default CreateUser;
