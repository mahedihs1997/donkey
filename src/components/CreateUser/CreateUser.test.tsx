import { MockedProvider } from '@apollo/react-testing';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import CreateUser from '.';
import { CREATE_USER_MUTATION } from '@domain/mutations/user';

const mocks = [
  {
    request: {
      query: CREATE_USER_MUTATION,
      variables: {
        username: 'Ugendo@buugendo.com',
        password: '123',
        name: 'Sickendo',
      },
    },
    result: {
      data: {
        token: 'dont-look-bro',
      },
    },
  },
];

describe('<CreateUser />', () => {
  it('should render without throwing an error and match snapshot', () => {
    const testRenderer = TestRenderer.create(
      <MockedProvider addTypename={false} mocks={mocks}>
        <CreateUser
          onSuccess={jest.fn()}
          open={false}
          handleClose={jest.fn()}
        />
      </MockedProvider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
