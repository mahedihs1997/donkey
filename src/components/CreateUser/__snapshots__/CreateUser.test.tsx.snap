// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateUser /> should render without throwing an error and match snapshot 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="makeStyles-form-1"
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Name
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="name"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
      
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Username
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="username"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
      
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Password
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="password"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
      
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Create New User
      </span>
    </button>
  </div>
</form>
`;
